examples/tv-casting-app/android/App/app/src/main/jni/com/chip/casting/TvCastingApp.java

```diff
@@ -109,50 +120,62 @@ public boolean initApp(Context applicationContext, AppParameters appParameters)
-
-    WifiManager wifiManager =
-        (WifiManager) applicationContext.getSystemService(Context.WIFI_SERVICE);
-    multicastLock = wifiManager.createMulticastLock("multicastLock");
-    multicastLock.setReferenceCounted(true);
-    multicastLock.acquire();
-
-    nsdManager = (NsdManager) applicationContext.getSystemService(Context.NSD_SERVICE);
-    nsdDiscoveryListener =
-        new NsdDiscoveryListener(
-            nsdManager,
-            DISCOVERY_TARGET_SERVICE_TYPE,
-            DISCOVERY_TARGET_DEVICE_TYPE_FILTER,
-            preCommissionedVideoPlayers,
-            discoverySuccessCallback,
-            discoveryFailureCallback,
-            nsdManagerResolverAvailState);
-
-    nsdManager.discoverServices(
-        DISCOVERY_TARGET_SERVICE_TYPE, NsdManager.PROTOCOL_DNS_SD, nsdDiscoveryListener);
-    Log.d(TAG, "TvCastingApp.discoverVideoPlayerCommissioners started");
-    this.discoveryStarted = true;
+      List<VideoPlayer> preCommissionedVideoPlayers = readCachedVideoPlayers();
+
+      WifiManager wifiManager =
+          (WifiManager) applicationContext.getSystemService(Context.WIFI_SERVICE);
+      multicastLock = wifiManager.createMulticastLock("multicastLock");
+      multicastLock.setReferenceCounted(true);
+      multicastLock.acquire();
+
+      nsdManager = (NsdManager) applicationContext.getSystemService(Context.NSD_SERVICE);
+      nsdDiscoveryListener =
+          new NsdDiscoveryListener(
+              nsdManager,
+              DISCOVERY_TARGET_SERVICE_TYPE,
+              DISCOVERY_TARGET_DEVICE_TYPE_FILTER,
+              preCommissionedVideoPlayers,
+              discoverySuccessCallback,
+              discoveryFailureCallback,
+              nsdManagerResolverAvailState);
+
+      nsdManager.discoverServices(
+          DISCOVERY_TARGET_SERVICE_TYPE, NsdManager.PROTOCOL_DNS_SD, nsdDiscoveryListener);
+      Log.d(TAG, "TvCastingApp.discoverVideoPlayerCommissioners started");
+      this.discoveryStarted = true;
+    }
   }
 
   public void stopVideoPlayerDiscovery() {
-    Log.d(TAG, "TvCastingApp trying to stop video player discovery");
-    if (this.discoveryStarted
-        && nsdManager != null
-        && multicastLock != null
-        && nsdDiscoveryListener != null) {
-      Log.d(TAG, "TvCastingApp stopping Video Player commissioner discovery");
-      nsdManager.stopServiceDiscovery(nsdDiscoveryListener);
-      if (multicastLock.isHeld()) {
-        multicastLock.release();
+    synchronized (discoveryLock) {
+      Log.d(TAG, "TvCastingApp trying to stop video player discovery");
+      if (this.discoveryStarted
+          && nsdManager != null
+          && multicastLock != null
+          && nsdDiscoveryListener != null) {
+        Log.d(TAG, "TvCastingApp stopping Video Player commissioner discovery");
+        try {
+          nsdManager.stopServiceDiscovery(nsdDiscoveryListener);
+        } catch (IllegalArgumentException e) {
+          Log.e(
```

Comment:
maybe log a warn since we're continuing, an error implies something really bad happened

Reply:
Done